spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http/RestClientEurekaHttpClient.java

```diff
@@ -0,0 +1,205 @@
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.netflix.eureka.http;
+
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import com.netflix.appinfo.InstanceInfo;
+import com.netflix.appinfo.InstanceInfo.InstanceStatus;
+import com.netflix.discovery.shared.Application;
+import com.netflix.discovery.shared.Applications;
+import com.netflix.discovery.shared.transport.EurekaHttpClient;
+import com.netflix.discovery.shared.transport.EurekaHttpResponse;
+import com.netflix.discovery.shared.transport.EurekaHttpResponse.EurekaHttpResponseBuilder;
+import com.netflix.discovery.util.StringUtil;
+
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.client.RestClient;
+import org.springframework.web.util.UriBuilder;
+
+import static com.netflix.discovery.shared.transport.EurekaHttpResponse.anEurekaHttpResponse;
+
+/**
+ * @author Wonchul Heo
+ */
+public class RestClientEurekaHttpClient implements EurekaHttpClient {
+
+	private final RestClient restClient;
+
+	public RestClientEurekaHttpClient(RestClient restClient) {
+		this.restClient = restClient;
+	}
+
+	@Override
+	public EurekaHttpResponse<Void> register(InstanceInfo info) {
+		final ResponseEntity<Void> response = restClient.post()
+				.uri(builder -> builder.pathSegment("apps", info.getAppName()).build()).body(info)
+				.header(HttpHeaders.ACCEPT_ENCODING, "gzip")
+				.header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE).retrieve().toBodilessEntity();
+		return eurekaHttpResponse(response);
+	}
+
+	@Override
+	public EurekaHttpResponse<Void> cancel(String appName, String id) {
+		final ResponseEntity<Void> response = restClient.delete()
+				.uri(builder -> builder.pathSegment("apps", appName, id).build()).retrieve().toBodilessEntity();
+		return eurekaHttpResponse(response);
+	}
+
+	@Override
+	public EurekaHttpResponse<InstanceInfo> sendHeartBeat(String appName, String id, InstanceInfo info,
+			InstanceStatus overriddenStatus) {
+		final Function<UriBuilder, URI> uriFunction = builder -> builder.pathSegment("apps", appName, id)
+				.queryParam("status", info.getStatus().toString())
+				.queryParam("lastDirtyTimestamp", info.getLastDirtyTimestamp().toString()).queryParamIfPresent(
+						"overriddenstatus", Optional.ofNullable(overriddenStatus).map(InstanceStatus::name))
+				.build();
```

Comment:
I used `UriFunction` unlike other classes. WDYT? 🤔 

See https://github.com/spring-cloud/spring-cloud-netflix/blob/f21d0046cdbecf9535673ad13b7451818ac29995/spring-cloud-netflix-eureka-client/src/main/java/org/springframework/cloud/netflix/eureka/http /WebClientEurekaHttpClient.java#L73...L75

Reply:
Related https://github.com/spring-cloud/spring-cloud-netflix/pull/4163, so I guess it should be accepted as there's a linked bug report.

Reply:
Seems like a good idea.